### PR TITLE
JDK21 adds PreviewFeature APIs and JVM methods

### PR DESCRIPTION
**Pull request by JasonFengJ9**
`JDK21` adds `PreviewFeature` APIs and JVM methods

Added the following PreviewFeature APIs:
```
stringConcatCoder(char value)
stringBuilderConcatMix(long lengthCoder, StringBuilder sb)
stringBuilderConcatPrepend(long lengthCoder, byte[] buf, StringBuilder
sb)
```
Also added `JVM_PrintWarningAtDynamicAgentLoad`.

Rebased on 
* https://github.com/eclipse-openj9/openj9/pull/17125

required by
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: jcl/src/java.base/share/classes/java/lang/Access.java [hunk 1]
```diff
@@ -600,6 +603,21 @@ public char getUTF16Char(byte[] val, int index) {
 	public int countPositives(byte[] ba, int off, int len) {
 		return StringCoding.countPositives(ba, off, len);
 	}
+
+	@PreviewFeature(feature = PreviewFeature.Feature.STRING_TEMPLATES)
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/java/lang/Access.java:607]: What's the point of the annotation on this and the next two methods?

File: runtime/redirector/forwarders.m4 [hunk 2]
```diff
@@ -411,6 +411,10 @@ _IF([JAVA_SPEC_VERSION >= 20],
 	[_X(JVM_VirtualThreadHideFrames, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide)])
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_IsForeignLinkerSupported, JNICALL, false, jboolean, void)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadStart, JNICALL, false, void, JNIEnv *env, jobject vthread)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadEnd, JNICALL, false, void, JNIEnv *env, jobject vthread)])
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_VirtualThreadMount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean firstMount)])
 _IF([JAVA_SPEC_VERSION >= 21],
```
Comment (keithc-ca) [inline runtime/redirector/forwarders.m4:420]: Please keep the entries for Java 21 sorted by name (like in `exports.cmake`).

File: runtime/j9vm/javanextvmi.cpp [hunk 3]
```diff
@@ -507,6 +507,18 @@ JVM_VirtualThreadUnmount(JNIEnv* env, jobject vthread, jboolean hide, jboolean l
 	}
 }
 
+JNIEXPORT void JNICALL
+JVM_VirtualThreadStart(JNIEnv* env, jobject vthread)
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:511]: Space before, not after `*`; here and for `JVM_VirtualThreadEnd`.

File: runtime/j9vm/javanextvmi.cpp [hunk 4]
```diff
@@ -507,6 +507,18 @@ JVM_VirtualThreadUnmount(JNIEnv* env, jobject vthread, jboolean hide, jboolean l
 	}
 }
 
+JNIEXPORT void JNICALL
```
Comment (keithc-ca) [inline runtime/j9vm/javanextvmi.cpp:510]: The signatures of `JVM_VirtualThreadMount()` and `JVM_VirtualThreadUnmount()` have changed; see https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/master/src/hotspot/share/include/jvm.h#L1152-L1156.

File: runtime/redirector/forwarders.m4 [hunk 5]
```diff
@@ -411,6 +411,10 @@ _IF([JAVA_SPEC_VERSION >= 20],
 	[_X(JVM_VirtualThreadHideFrames, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide)])
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_IsForeignLinkerSupported, JNICALL, false, jboolean, void)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadStart, JNICALL, false, void, JNIEnv *env, jobject vthread)])
+_IF([JAVA_SPEC_VERSION >= 21],
+	[_X(JVM_VirtualThreadEnd, JNICALL, false, void, JNIEnv *env, jobject vthread)])
 _IF([JAVA_SPEC_VERSION >= 21],
 	[_X(JVM_VirtualThreadMount, JNICALL, false, void, JNIEnv *env, jobject vthread, jboolean hide, jboolean firstMount)])
```
Comment (keithc-ca) [inline runtime/redirector/forwarders.m4:419]: The final parameter of `JVM_VirtualThreadMount()` and `JVM_VirtualThreadUnmount()` has been removed.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

**Comment by babsingh** [comment]
#17125 addresses all issues related to the `JVM_VirtualThread*` functions. Instead of adding stubs, #17125 can be used to unblock the `jdk-next/openj9-staging` build. Also, adding stubs will cause merge conflicts in #17125. fyi @fengxue-IS @thallium

**Comment by JasonFengJ9** [comment]
> https://github.com/eclipse-openj9/openj9/pull/17125 can be used to unblock the jdk-next/openj9-staging build

If https://github.com/eclipse-openj9/openj9/pull/17125 has a better chance to be merged first, this PR can just keep j.l.Access new APIs.

**Comment by babsingh** [comment]
`sanity.functional` test suite, which is used to promote `openj9-staging` builds, has `Jep425Tests_testVirtualThread`. All the changes from https://github.com/eclipse-openj9/openj9/pull/17125 are probably needed to pass these tests. Stubs will lead to test failures.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 1 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: This is prepare for [JEP 430: String Templates (Preview)](https://openjdk.org/jeps/430)

**Comment by thallium** [comment]
> @thallium Can you confirm if #17125 is ready to be merged?

@babsingh Yes.

File: runtime/redirector/forwarders.m4 (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/redirector/forwarders.m4]: Sorted

File: runtime/j9vm/javanextvmi.cpp (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/j9vm/javanextvmi.cpp]: Adjusted

File: runtime/redirector/forwarders.m4 (hunk 5 above)
Comment (JasonFengJ9) [reply runtime/redirector/forwarders.m4:417]: I think this signature change is implemented by 
* https://github.com/eclipse-openj9/openj9/pull/17125

**Comment by JasonFengJ9** [comment]
Just updated this PR to address the comments except for the actual implementation for `JVM_VirtualThread[Mount|Unmount]`.

This PR probably is not sufficient for `openj9-staging` promotion though it could enable [Merge-to-Staging-JDKNext](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJDK%20Team%20-%20Jobs/job/Merge%20Master%20to%20openj9-staging%20Jobs/job/Merge-to-Staging-JDKNext/) for further openjdk content consumption before https://github.com/eclipse-openj9/openj9/pull/17125 is to be merged.

File: jcl/src/java.base/share/classes/java/lang/Access.java (hunk 1 above)
Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: They have meaning for the interface definition; but I don't see any value here. Please explain why the annotations are necessary.

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: This is to match openjdk implementation.
I think `javac` processes it - https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/31c58a953edb55b0a3a4fe5fbae4ff17f8eb8810/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java#L343C3-L344

Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: I expect that only applies to code that _uses_ string templates, and isn't relevant to our implementation of JavaLangAccess. We'll only have to remove the annotations later (once the feature is out of preview); I don't see any value to adding them here in the first place.

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: The three `JavaLangAccess` APIs annotated with `STRING_TEMPLATES` are openjdk `StringBuilder` usages, I expect openjdk uses string templates in those cases assuming the annotation was added for such purpose.

Comment (keithc-ca) [reply jcl/src/java.base/share/classes/java/lang/Access.java:607]: The only uses I see (so far) are in `java.util.FormatItem`; only those uses are affected by the annotations on the interface methods; they have no visibility of the implementations.

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/Access.java]: Removed `@PreviewFeature` annotation from `Access` implementation.

Also added `JVM_PrintWarningAtDynamicAgentLoad()` temporary implementation which was just introduced by latest `jdk-21+25`.

File: runtime/redirector/forwarders.m4 (hunk 5 above)
Comment (keithc-ca) [reply runtime/redirector/forwarders.m4:419]: Please remove those parameters; #17125 should not conflict with this no matter which order the two pull requests are merged.

Comment (JasonFengJ9) [reply runtime/redirector/forwarders.m4]: Removed `firstMount` in `JVM_VirtualThreadMount()`, `lastUnmount` in `JVM_VirtualThreadUnmount`, and passed `JNI_TRUE` to `JVM_VirtualThreadMount[Begin|End]` and `JVM_VirtualThreadUnmount[Begin|End]`, these are temporary implementation, expected to be replaced by https://github.com/eclipse-openj9/openj9/pull/17125.

File: runtime/j9vm/javanextvmi.cpp [hunk 6]
```diff
@@ -487,6 +487,15 @@ JVM_VirtualThreadHideFrames(JNIEnv *env, jobject vthread, jboolean hide)
...
+JVM_PrintWarningAtDynamicAgentLoad()
+{
+	/* A temporary implementation, an actual solution is provided via
+	 * https://github.com/eclipse-openj9/openj9/issues/17500
+	 */
+	return JNI_TRUE;
+}
+
 JNIEXPORT void JNICALL
 JVM_VirtualThreadMount(JNIEnv* env, jobject vthread, jboolean hide, jboolean firstMount)
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/j9vm/javanextvmi.cpp]: Removed both https://github.com/eclipse-openj9/openj9/pull/17514#discussion_r1224553801

**Comment by keithc-ca** [comment]
Please rebase this on https://github.com/eclipse-openj9/openj9/pull/17125; see https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608#issuecomment-1589778453.

**Comment by JasonFengJ9** [comment]
Rebased on https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608#issuecomment-1589778453

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Tested in https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/5 and https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/608.

**Pull request merged**